classes/local/accredible.php

```diff
@@ -0,0 +1,94 @@
+ *
+ * @package    mod_accredible
+ * @subpackage accredible
+ * @copyright  Accredible <dev@accredible.com>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class accredible {
+
+    /**
+     * Saves or updates an Accredible plugin instance record in the 'accredible' table.
+     * This function handles both the creation of new records and the updating of existing ones.
+     *
+     * @param stdClass $post An object containing the incoming data from the form submission.
+     * @param stdClass|null $existingrecord Optional. Existing certificate data to be updated.
+     * @return bool|int Returns the new record ID on insert, or true on update success.
+     */
+    public function save_record($post, $existingrecord = null) {
+        global $DB;
+
+        $dbrecord = new \stdClass();
+        $dbrecord->completionactivities = isset($post->completionactivities) ? $post->completionactivities : null;
+        $dbrecord->name = $post->name;
+        $dbrecord->finalquiz = $post->finalquiz;
+        $dbrecord->passinggrade = $post->passinggrade;
+        $dbrecord->includegradeattribute = isset($post->includegradeattribute) ? $post->includegradeattribute : 0;
+        $dbrecord->gradeattributegradeitemid = $post->gradeattributegradeitemid;
+        $dbrecord->gradeattributekeyname = $post->gradeattributekeyname;
+        $dbrecord->groupid = $post->groupid;
+        $dbrecord->attributemapping = $this->build_attribute_mapping_list($post);
+        $dbrecord->timecreated = time();
+
+        if ($existingrecord) {
+            $dbrecord->id = $post->instance;
+
+            if ($existingrecord->achievementid) {
+                $dbrecord->certificatename = $post->certificatename;
+                $dbrecord->description = $post->description;
+                $dbrecord->achievementid = $post->achievementid;
+                // Keep the existing groupid.
+                $dbrecord->groupid = $existingrecord->groupid;
+            } else {
+                $dbrecord->course = $post->course;
+            }
+
+            return $DB->update_record('accredible', $dbrecord);
+        } else {
+            $dbrecord->course = $post->course;
+            return $DB->insert_record('accredible', $dbrecord);
+        }
+    }
+
+    /**
+     * Builds a JSON encoded attribute mapping list to be stored in the DB based on the provided post data.
+     *
+     * @param object $post The post data containing the course field mappings, course custom field mappings,
+     * and user field mappings.
+     * @return string JSON encoded attribute mapping list.
+     */
+    private function build_attribute_mapping_list($post) {
+        // Combine all the mappings into a single array. Expects empty arrays if no mappings are present.
+        $mergedmappings = array_merge($post->coursefieldmapping, $post->coursecustomfieldmapping, $post->userfieldmapping);
```

Comment:
@kei178 In this function, I'm expecting empty arrays if no mappings are present (that's why I left that comment there). As such, I'm expecting the form submission code to always return empty arrays instead of `null`. 

Let me know what you think. I can add null validation here if you think it's a better approach.